### PR TITLE
Add Frozen Treasure Highlighter

### PR DESCRIPTION
**Pull request by heyngra**
1. It uses the previously used in https://github.com/NotEnoughUpdates/NotEnoughUpdates/pull/266 system of checking if you can see the block.
2. There's an [NullPointerException](https://hst.sh/raw/pizizunofu) which occurs, pretty sure it's not related with this commit (it has been apperaing there even when disabling any events in the class itself), which also doesn't give any traceback to currently existing class in NEU

### CODE REVIEW

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+	@Override
+	protected boolean isValidHighlightSpot(BlockPos key) {
+		World w = Minecraft.getMinecraft().theWorld;
+		if (w == null) return false;
+		Block b = w.getBlockState(key).getBlock();
+		return b == Blocks.ice;
+	}
+
+	protected static boolean canSee(MovingObjectPosition hitResult, BlockPos bp) {
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:59]: this method seems unused?

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	public void onWorldRenderLast(RenderWorldLastEvent event) {
+		if (!isEnabled()) return;
+		World w = Minecraft.getMinecraft().theWorld;
+		if (w == null) return;
+		List<Entity> entities = w.getLoadedEntityList();
+		for (Entity e : entities) {
+			if (!(e instanceof EntityArmorStand) || ((EntityArmorStand) e).getCurrentArmor(3)==null) continue;
+			BlockPos blockPos = e.getPosition();
+			BlockPos blockPos2 = blockPos.add(0, 1, 0);
+			if (tryRegisterInterest(blockPos2)) {
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:69]: i dont think you should do this in every render event, and instead do it in a tick

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+				RenderUtils.renderBoundingBox(blockPos2, getColor(blockPos2), event.partialTicks);
+			}
+		}
+	}
+	@Override
+	protected int getColor(BlockPos blockPos) {
+		return SpecialColour.specialToChromaRGB(NotEnoughUpdates.INSTANCE.config.world.frozenTreasuresColor);
+	}
+	@Override
+	public boolean tryRegisterInterest(double x, double y, double z) {
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java]: I don't really see a point overwriting this method, the superclass basically does the same

Comment (heyngra) [reply src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java]: https://github.com/NotEnoughUpdates/NotEnoughUpdates/pull/477/commits/161515b1002e7eb04277f27fe0c77fdc3487e63a

**Comment by heyngra** [comment]
uh, re-requesting a review shouldn't remove other requests

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		}
+	}
+	@SubscribeEvent
+	public void onWorldRenderLast(RenderWorldLastEvent event) {
+		if (!isEnabled()) return;
+		World w = Minecraft.getMinecraft().theWorld;
+		if (w == null) return;
+		for (Entity e : highlightedTreasures) {
+			BlockPos blockPos = e.getPosition().add(0, 1, 0);
+			if (tryRegisterInterest(blockPos)) {
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:81]: Checking for interest should be done in ontick (which would also avoid overwriting this method at all)

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			}
+		}
+	}
+
+	@Override
+	protected int getColor(BlockPos blockPos) {
+		return SpecialColour.specialToChromaRGB(NotEnoughUpdates.INSTANCE.config.world.frozenTreasuresColor);
+	}
+
+	public boolean tryRegisterInterest(BlockPos blockPos) {
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:92]: If you must keep this method (as i dont think you should, see other comments), maybe rename it to something more accurate.

**Review by romangraef — CHANGES_REQUESTED** [review]

**Review by heyngra — COMMENTED** [review]
not sure if it's name is accurate enough

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	protected boolean isValidHighlightSpot(BlockPos key) {
+		World w = Minecraft.getMinecraft().theWorld;
+		if (w == null) return false;
+		Block b = w.getBlockState(key).getBlock();
+		return b == Blocks.ice;
+	}
+
+	@SubscribeEvent
+	public void onTickNew(TickEvent.ClientTickEvent event) {
+		if (event.phase != TickEvent.Phase.END||!isEnabled()) return;
```
Comment (hannibal002) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:59]: Please reformat the file. (CONTROL+ALT+L in IntellIJ)

**Review by hannibal002 — COMMENTED** [review]
Looging good to me.

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java (hunk 6 above)
Comment (heyngra) [reply src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java]: I didn't know about this shortcut! Thanks!

**Review by hannibal002 — DISMISSED** [review]

**Review by NopoTheGamer — APPROVED** [review]

**Review by NopoTheGamer — CHANGES_REQUESTED** [review]
nvm crashed when closing the game https://hst.sh/raw/yuwanoweru

File: src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+		if (w == null) return false;
+		Block b = w.getBlockState(key).getBlock();
+		return b == Blocks.ice;
+	}
+
+	@SubscribeEvent
+	public void onTickNew(TickEvent.ClientTickEvent event) {
+		if (event.phase != TickEvent.Phase.END || !isEnabled()) return;
+
+		List<Entity> entities = Minecraft.getMinecraft().theWorld.getLoadedEntityList();
```
Comment (hannibal002) [inline src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:61]: theWorld can be null.

Comment (heyngra) [reply src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/FrozenTreasuresHighlighter.java:61]: aight

**Review by NopoTheGamer — APPROVED** [review]

**Review by romangraef — APPROVED** [review]

**Comment by heyngra** [comment]
tip for maybe someone in the future: forge is stupid and for some reason helmet is id 3 and not 0

**Comment by romangraef** [comment]
@heyngra oh actually, that reminds me: did you test with sba? sba changes the slot ids of armor sometimes

**Comment by romangraef** [comment]
Nvm, that was only for players, not for armor stands

**Pull request merged**